### DISCUSSION TITLE
Can't find a sniffer that removes extra (more than 1) blank lines

### DISCUSSION THREAD

**Discussion by ghost** [PHPCSStandards/PHP_CodeSniffer#575] [Q&A] [answered]
Hi, I have both PHP_CodeSniffer and PHPCSExtra installed and I can't find a sniffer (or sniffers that when combined give that effect) that removes extra blank lines.

Example:
```diff
function foo()
{
    echo 'one';

-                  
    echo 'two';
}
```

**Comment by jrfnl** [comment]
Try `Squiz.WhiteSpace.SuperfluousWhitespace`

**Reply by ghost** [reply]
Tried that. Doesn't help :(

**Reply by ghost** [reply]
However the doc says it should work `vendor/bin/phpcs --generator=Text --standard=Squiz --sniffs=Squiz.WhiteSpace.SuperfluousWhitespace` (btw. if it worked then would it work even outside a function?)

```
There should be no consecutive blank lines in functions.

----------------------------------------- CODE COMPARISON ------------------------------------------
| Valid: Functions do not contain multiple empty | Invalid: Functions contain multiple empty       |
| lines in a row.                                | lines in a row.                                 |
----------------------------------------------------------------------------------------------------
| function myFunction()                          | function myFunction()                           |
| {                                              | {                                               |
|     echo 'code here';                          |     echo 'code here';                           |
|                                                |                                                 |
|     echo 'code here';                          |                                                 |
| }                                              |     echo 'code here';                           |
|                                                | }                                               |
----------------------------------------------------------------------------------------------------

```

**Reply by ghost** [reply]
Hmmm, I've commented out all other rules in config file and it worked (this one was the last one in the file). Wondering why :thinking:

**Reply by ghost** [reply]
Probably because PSR12 ruleset has this:
```xml
    <rule ref="Squiz.WhiteSpace.SuperfluousWhitespace.EmptyLines">
        <severity>0</severity>
    </rule>
```

**Reply by jrfnl** [reply, marked as answer]
I wouldn't have mentioned it if it didn't...

> Hmmm, I've commented out all other rules in config file and it worked (this one was the last one in the file). Wondering why

You're probably including a predefined ruleset which already includes that sniff, but excludes the `Squiz.WhiteSpace.SuperfluousWhitespace.EmptyLines` error code.

Revert whatever changes you made in your ruleset now and try adding the below towards the end instead:
```xml
<rule ref="Squiz.WhiteSpace.SuperfluousWhitespace.EmptyLines">
    <severity>5</severity>
</rule>
```

That should overrule the earlier exclusion.

**Reply by ghost** [reply]
@jrfnl I wonder:
- why my line `<rule ref="Squiz.WhiteSpace.SuperfluousWhitespace"/>` doesn't reset the one from PSR12 ruleset that is being imported earlier. It feel like it should reset the whole rule, i.e. also the specific error identifiers (EmptyLines, EndFile, StartFIle) which are a subset of that rule so when the root rule is overwritten then it should reset everything 
- where can I find this identifiers after last dot (EmptyLines, StartFile, EmptyLines)? They're not in Sniff file nor in the `vendor/bin/phpcs --generator=Text --standard=Squiz --sniffs=Squiz.WhiteSpace.SuperfluousWhitespace` output. (actually they are but somewhere deep inside `process()` method the implementation)

**Reply by ghost** [reply]
BTW. This sniff doesn't work outside a method/function

Example:
```
class Foo
{
    use MyTrait;


    private const MY_CONST = 1;
}
```

**Reply by jrfnl** [reply]
> where can I find this identifiers after last dot (EmptyLines, StartFile, EmptyLines)?

Run with the `-s` flag and you get to see the error codes.

> They're not in Sniff file

They are, as that's where they are coming from.

https://github.com/PHPCSStandards/PHP_CodeSniffer/blob/329ad21d895bb2c4b7aacdb2b70b3c938a855823/src/Standards/Squiz/Sniffs/WhiteSpace/SuperfluousWhitespaceSniff.php#L245-L246

> why my line <rule ref="Squiz.WhiteSpace.SuperfluousWhitespace"/> doesn't reset the one from PSR12 ruleset that is being imported earlier. It feel like it should reset the whole rule, i.e. also the specific error identifiers (EmptyLines, EndFile, StartFIle) which are a subset of that rule so when the root rule is overwritten then it should reset everything

That's something historical. Changing it, however, would be a breaking change, so can't be done easily.
I know there is a change in PHPCS 4.0 in how the ruleset gets read, but I don't know for sure if that would change the behaviour you are experiencing.

**Reply by jrfnl** [reply]
> BTW. This sniff doesn't work outside a method/function

Correct and that's by design. There are other sniffs which have (configurable) rules about number of blank lines between functions, properties, use statements etc.

Having one "catch-all" sniff for blank lines would very quickly create conflict with other sniffs, especially when auto-fixing, so having these rules in different sniffs, each with their own concern helps prevent that.

**Reply by ghost** [reply]
Thank you. Will dig further to find a desired one.

However I think PHP_CodeSniffer and PHP-CS-Fixer should enter into cooperation and maintain some kind of table that lists what rule (or set of rules) gives the same result in both tools. IMO exploring and learning another CS tool feels like learning to walk again. I think the life is beautiful enough that it's not worth spending time on such things.

P.S. It's not that I want to switch from one tool to another. I have to because my current company uses PHP_CodeSniffer but their config is very poor so I wanted to improve it but it has already been taking me too long :/

**Reply by jrfnl** [reply]
You are welcome to volunteer to create such a table and publish it somewhere.
